execution/gethexec/express_lane_service.go

```diff
@@ -0,0 +1,323 @@
+	auctionClosingDuration := time.Duration(roundTimingInfo.AuctionClosingSeconds) * time.Second
+	return &expressLaneService{
+		auctionContract:          auctionContract,
+		chainConfig:              chainConfig,
+		initialTimestamp:         initialTimestamp,
+		auctionClosing:           auctionClosingDuration,
+		roundControl:             lru.NewBasicLRU[uint64, *expressLaneControl](8), // Keep 8 rounds cached.
+		auctionContractAddr:      auctionContractAddr,
+		roundDuration:            roundDuration,
+		seqClient:                sequencerClient,
+		logs:                     make(chan []*types.Log, 10_000),
+		messagesBySequenceNumber: make(map[uint64]*timeboost.ExpressLaneSubmission),
+	}, nil
+}
+
+func (es *expressLaneService) Start(ctxIn context.Context) {
+	es.StopWaiter.Start(ctxIn, es)
+
+	// Log every new express lane auction round.
+	es.LaunchThread(func(ctx context.Context) {
+		log.Info("Watching for new express lane rounds")
+		now := time.Now()
+		waitTime := es.roundDuration - time.Duration(now.Second())*time.Second - time.Duration(now.Nanosecond())
+		time.Sleep(waitTime)
+		ticker := time.NewTicker(time.Minute)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case t := <-ticker.C:
+				round := timeboost.CurrentRound(es.initialTimestamp, es.roundDuration)
+				log.Info(
+					"New express lane auction round",
+					"round", round,
+					"timestamp", t,
+				)
+				es.Lock()
+				// Reset the sequence numbers map for the new round.
+				es.messagesBySequenceNumber = make(map[uint64]*timeboost.ExpressLaneSubmission)
+				es.Unlock()
+			}
+		}
+	})
+	es.LaunchThread(func(ctx context.Context) {
+		log.Info("Monitoring express lane auction contract")
+		// Monitor for auction resolutions from the auction manager smart contract
+		// and set the express lane controller for the upcoming round accordingly.
+		latestBlock, err := es.seqClient.HeaderByNumber(ctx, nil)
+		if err != nil {
+			// TODO: Should not be a crit.
+			log.Crit("Could not get latest header", "err", err)
+		}
+		fromBlock := latestBlock.Number.Uint64()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(time.Millisecond * 250):
+				latestBlock, err := es.seqClient.HeaderByNumber(ctx, nil)
+				if err != nil {
+					log.Crit("Could not get latest header", "err", err)
+				}
+				toBlock := latestBlock.Number.Uint64()
+				if fromBlock == toBlock {
+					continue
+				}
+				filterOpts := &bind.FilterOpts{
+					Context: ctx,
+					Start:   fromBlock,
+					End:     &toBlock,
+				}
+				it, err := es.auctionContract.FilterAuctionResolved(filterOpts, nil, nil, nil)
+				if err != nil {
+					log.Error("Could not filter auction resolutions event", "error", err)
+					continue
+				}
+				for it.Next() {
+					log.Info(
+						"New express lane controller assigned",
+						"round", it.Event.Round,
+						"controller", it.Event.FirstPriceExpressLaneController,
+					)
+					es.Lock()
```

Comment:
I think locking deserves some overview.
Would suggest trying to split between a lock used while processing incoming transaction in sequenceExpressLaneSubmission (which writes the message sequence and only reads round info of current round) vs lock used while while processing new round info (which doesn't care about incoming transactions and writes - hopefully only next-round). If needed, incoming-Tx may hold biefly the round lock just to check current round info

Reply:
Thanks for pointing this out. We can take advantage of roundControl's thread-safety (via lru.Cache) to avoid holding the full lock while just reading round info. 

I've removed the unnecessary lock in currentRoundHasController and split the locking in sequenceExpressLaneSubmission to only lock during transaction processing.

This should reduce lock contention by separating round info reads from transaction processing. Let me know if you had something else in mind.

I made the fix in https://github.com/OffchainLabs/nitro/pull/2787/files